File: app/src/main/java/com/jeanbarrossilva/power/LengthFragment.java
package com.jeanbarrossilva.power;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;

import android.view.LayoutInflater;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.PopupMenu;
import android.widget.TextView;

import org.jetbrains.annotations.NotNull;

import static com.jeanbarrossilva.power.MainActivity.DEFAULT_BOUNCE_IN_SETTING;

public class LengthFragment extends CalculatorFragment {
    private TextView inputSymbol;

    private TextView conversionResult;
    private TextView conversionSymbolResult;

    private Button unit;
    private Button[] options =  new Button[10];

    private Button lightYear;
    private Button kilometer;
    private Button hectometer;
    private Button decameter;
    private Button mile;
    private Button meter;
    private Button decimeter;
    private Button centimeter;
    private Button millimeter;
    private Button micrometer;

    public LengthFragment() {

    }

    @Override
    public void onAttach(@NotNull Context context) {
        super.onAttach(context);
        this.context = context;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_calculator_length, container, false);
        mainActivity = (MainActivity) getActivity();

        input = view.findViewById(R.id.input);
        inputSymbol = view.findViewById(R.id.input_symbol);

        unit = view.findViewById(R.id.unit);

        conversionResult = view.findViewById(R.id.option_conversion_number_result);
        conversionSymbolResult = view.findViewById(R.id.option_conversion_symbol_result);

        othersHorizontalScrollView = view.findViewById(R.id.others_horizontal_scroll_view);
        othersHorizontalScrollView.setHorizontalScrollBarEnabled(false);

        lightYear = view.findViewById(R.id.light_year);
        kilometer = view.findViewById(R.id.kilometer);
        hectometer = view.findViewById(R.id.hectometer);
        decameter = view.findViewById(R.id.decameter);
        mile = view.findViewById(R.id.mile);
        meter = view.findViewById(R.id.meter);
        decimeter = view.findViewById(R.id.decimeter);
        centimeter = view.findViewById(R.id.centimeter);
        millimeter = view.findViewById(R.id.millimeter);
        micrometer = view.findViewById(R.id.micrometer);

        options[0] = lightYear;
        options[1] = kilometer;
        options[2] = hectometer;
        options[3] = decameter;
        options[4] = mile;
        options[5] = meter;
        options[6] = decimeter;
        options[7] = centimeter;
        options[8] = millimeter;
        options[9] = micrometer;

        keypadButtons[10] = view.findViewById(R.id.decimal_separator);
        calculatorMode = view.findViewById(R.id.calculator_mode);
        delete = view.findViewById(R.id.delete);

        units();
        inputOption();

        // Default configuration (meter to kilometer).
        inputSymbol.setText(getString(R.string.meter_symbol));
        mainActivity.selectButton(meter, options, R.drawable.option_clicked);
        mainActivity.getPreferencesEditor().putString("convertFrom", "meter")
                .apply();

        unit.setText(getString(R.string.kilometer));
        conversionSymbolResult.setText(getString(R.string.kilometer_symbol));
        mainActivity.getPreferencesEditor().putString("convertTo", "kilometer")
                .apply();

        mainActivity.calculatorMode(context, calculatorMode);

        mainActivity.inputNumber(view, input, conversionResult, conversionSymbolResult, input.getText().toString());
        inputDecimalSeparator(input);
        mainActivity.delete(input, delete, conversionResult, conversionSymbolResult);

        return view;
    }

    @SuppressLint("ClickableViewAccessibility")
    private void units() {
        unit.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        mainActivity.bounceIn(unit, DEFAULT_BOUNCE_IN_SETTING);
                        break;
                    case MotionEvent.ACTION_UP:
                        final PopupMenu units;
                        MenuInflater inflater;

                        units = new PopupMenu(context, unit);
                        inflater = units.getMenuInflater();

                        inflater.inflate(R.menu.units_length, units.getMenu());
                        units.show();

                        units.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
                            @Override
                            public boolean onMenuItemClick(MenuItem item) {
                                if (item.getTitle().equals(getString(R.string.light_year))) {
                                    mainActivity.getPreferencesEditor().putString("convertTo", "lightYear")
                                            .apply();

                                    conversionSymbolResult.setText(getString(R.string.light_year_symbol));
                                } else if (item.getTitle().equals(getString(R.string.kilometer))) {
                                    mainActivity.getPreferencesEditor().putString("convertTo", "kilometer")
                                            .apply();

                                    conversionSymbolResult.setText(getString(R.string.kilometer_symbol));
                                    unit.setText(getString(R.string.kilometer));
                                } else if (item.getTitle().equals(getString(R.string.hectometer))) {
                                    mainActivity.getPreferencesEditor().putString("convertTo", "hectometer")
                                            .apply();

                                    conversionSymbolResult.setText(getString(R.string.hectometer_symbol));
                                    unit.setText(getString(R.string.hectometer));
                                } else if (item.getTitle().equals(getString(R.string.decameter))) {
                                    mainActivity.getPreferencesEditor().putString("convertTo", "decameter")
                                            .apply();

                                    conversionSymbolResult.setText(getString(R.string.decameter_symbol));
                                    unit.setText(getString(R.string.decameter));
                                } else if (item.getTitle().equals(getString(R.string.mile))) {
                                    mainActivity.getPreferencesEditor().putString("convertTo", "mile")
                                            .apply();

                                    conversionSymbolResult.setText(getString(R.string.mile_symbol));
                                    unit.setText(getString(R.string.mile));
                                } else if (item.getTitle().equals(getString(R.string.meter))) {
                                    mainActivity.getPreferencesEditor().putString("convertTo", "meter")
                                            .apply();

                                    conversionSymbolResult.setText(getString(R.string.meter_symbol));
                                    unit.setText(getString(R.string.meter));
                                } else if (item.getTitle().equals(getString(R.string.mile))) {
                                    mainActivity.getPreferencesEditor().putString("convertTo", "mile")
                                            .apply();

                                    conversionSymbolResult.setText(getString(R.string.mile_symbol));
                                    unit.setText(getString(R.string.mile));
                                } else if (item.getTitle().equals(getString(R.string.decimeter))) {
                                    mainActivity.getPreferencesEditor().putString("convertTo", "decimeter")
                                            .apply();

                                    conversionSymbolResult.setText(getString(R.string.decimeter_symbol));
                                    unit.setText(getString(R.string.decimeter));
                                } else if (item.getTitle().equals(getString(R.string.centimeter))) {
                                    mainActivity.getPreferencesEditor().putString("convertTo", "centimeter")
                                            .apply();

                                    conversionSymbolResult.setText(getString(R.string.centimeter_symbol));
                                    unit.setText(getString(R.string.centimeter));
                                } else if (item.getTitle().equals(getString(R.string.millimeter))) {
                                    mainActivity.getPreferencesEditor().putString("convertTo", "millimeter")
                                            .apply();

                                    conversionSymbolResult.setText(getString(R.string.millimeter_symbol));
                                    unit.setText(getString(R.string.millimeter));
                                } else if (item.getTitle().equals(getString(R.string.micrometer))) {
                                    mainActivity.getPreferencesEditor().putString("convertTo", "micrometer")
                                            .apply();

                                    conversionSymbolResult.setText(getString(R.string.micrometer_symbol));
                                    unit.setText(getString(R.string.micrometer));
                                }

                                mainActivity.calc(input, conversionResult, conversionSymbolResult);

                                return true;
                            }
                        });

                        break;
                }

                return true;
            }
        });
    }

    @SuppressLint("ClickableViewAccessibility")
    private void inputOption() {
        lightYear.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        mainActivity.bounceIn(lightYear, DEFAULT_BOUNCE_IN_SETTING);
                        break;
                    case MotionEvent.ACTION_UP:
                        lightYear.startAnimation(mainActivity.getBounceOut());

                        mainActivity.selectButton(lightYear, options, R.drawable.option_clicked);
                        inputSymbol.setText(getString(R.string.light_year_symbol));

                        mainActivity.getPreferencesEditor().putString("convertFrom", "lightYear")
                                .apply();
                        break;
                }

                mainActivity.calc(input, conversionResult, conversionSymbolResult);

                return true;
            }
        });

        kilometer.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        mainActivity.bounceIn(kilometer, DEFAULT_BOUNCE_IN_SETTING);
                        break;
                    case MotionEvent.ACTION_UP:
                        kilometer.startAnimation(mainActivity.getBounceOut());

                        mainActivity.selectButton(kilometer, options, R.drawable.option_clicked);
                        inputSymbol.setText(getString(R.string.kilometer_symbol));

                        mainActivity.getPreferencesEditor().putString("convertFrom", "kilometer")
                                .apply();
                        break;
                }

                mainActivity.calc(input, conversionResult, conversionSymbolResult);

                return true;
            }
        });

        hectometer.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        mainActivity.bounceIn(hectometer, DEFAULT_BOUNCE_IN_SETTING);
                        break;
                    case MotionEvent.ACTION_UP:
                        hectometer.startAnimation(mainActivity.getBounceOut());

                        mainActivity.selectButton(hectometer, options, R.drawable.option_clicked);
                        inputSymbol.setText(getString(R.string.hectometer_symbol));

                        mainActivity.getPreferencesEditor().putString("convertFrom", "hectometer")
                                .apply();
                        break;
                }

                mainActivity.calc(input, conversionResult, conversionSymbolResult);

                return true;
            }
        });

        decameter.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        mainActivity.bounceIn(decameter, DEFAULT_BOUNCE_IN_SETTING);
                        break;
                    case MotionEvent.ACTION_UP:
                        decameter.startAnimation(mainActivity.getBounceOut());

                        mainActivity.selectButton(decameter, options, R.drawable.option_clicked);
                        inputSymbol.setText(getString(R.string.decameter_symbol));

                        mainActivity.getPreferencesEditor().putString("convertFrom", "decameter")
                                .apply();
                        break;
                }

                mainActivity.calc(input, conversionResult, conversionSymbolResult);

                return true;
            }
        });

        mile.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        mainActivity.bounceIn(mile, DEFAULT_BOUNCE_IN_SETTING);
                        break;
                    case MotionEvent.ACTION_UP:
                        meter.startAnimation(mainActivity.getBounceOut());

                        mainActivity.selectButton(mile, options, R.drawable.option_clicked);
                        inputSymbol.setText(getString(R.string.mile_symbol));

                        mainActivity.getPreferencesEditor().putString("convertFrom", "mile")
                                .apply();
                        break;
                }

                mainActivity.calc(input, conversionResult, conversionSymbolResult);

                return true;
            }
        });

        meter.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        mainActivity.bounceIn(meter, DEFAULT_BOUNCE_IN_SETTING);
                        break;
                    case MotionEvent.ACTION_UP:
                        meter.startAnimation(mainActivity.getBounceOut());

                        mainActivity.selectButton(meter, options, R.drawable.option_clicked);
                        inputSymbol.setText(getString(R.string.meter_symbol));

                        mainActivity.getPreferencesEditor().putString("convertFrom", "meter")
                                .apply();
                        break;
                }

                mainActivity.calc(input, conversionResult, conversionSymbolResult);

                return true;
            }
        });

        decimeter.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        mainActivity.bounceIn(decimeter, DEFAULT_BOUNCE_IN_SETTING);
                        break;
                    case MotionEvent.ACTION_UP:
                        decimeter.startAnimation(mainActivity.getBounceOut());

                        mainActivity.selectButton(decimeter, options, R.drawable.option_clicked);
                        inputSymbol.setText(getString(R.string.decimeter_symbol));

                        mainActivity.getPreferencesEditor().putString("convertFrom", "decimeter")
                                .apply();
                        break;
                }

                mainActivity.calc(input, conversionResult, conversionSymbolResult);

                return true;
            }
        });

        centimeter.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        mainActivity.bounceIn(centimeter, DEFAULT_BOUNCE_IN_SETTING);
                        break;
                    case MotionEvent.ACTION_UP:
                        centimeter.startAnimation(mainActivity.getBounceOut());

                        mainActivity.selectButton(centimeter, options, R.drawable.option_clicked);
                        inputSymbol.setText(getString(R.string.centimeter_symbol));

                        mainActivity.getPreferencesEditor().putString("convertFrom", "centimeter")
                                .apply();
                        break;
                }

                mainActivity.calc(input, conversionResult, conversionSymbolResult);

                return true;
            }
        });

        millimeter.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        mainActivity.bounceIn(millimeter, DEFAULT_BOUNCE_IN_SETTING);
                        break;
                    case MotionEvent.ACTION_UP:
                        millimeter.startAnimation(mainActivity.getBounceOut());

                        mainActivity.selectButton(millimeter, options, R.drawable.option_clicked);
                        inputSymbol.setText(getString(R.string.millimeter_symbol));

                        mainActivity.getPreferencesEditor().putString("convertFrom", "millimeter")
                                .apply();
                        break;
                }

                mainActivity.calc(input, conversionResult, conversionSymbolResult);

                return true;
            }
        });

        micrometer.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        mainActivity.bounceIn(micrometer, DEFAULT_BOUNCE_IN_SETTING);
                        break;
                    case MotionEvent.ACTION_UP:
                        micrometer.startAnimation(mainActivity.getBounceOut());

                        mainActivity.selectButton(micrometer, options, R.drawable.option_clicked);
                        inputSymbol.setText(getString(R.string.micrometer_symbol));

                        mainActivity.getPreferencesEditor().putString("convertFrom", "micrometer")
                                .apply();
                        break;
                }

                mainActivity.calc(input, conversionResult, conversionSymbolResult);

                return true;
            }
        });
    }
}

File: app/src/main/java/com/jeanbarrossilva/power/LengthFragment.kt
package com.jeanbarrossilva.power

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.view.*
import android.widget.Button
import android.widget.HorizontalScrollView
import android.widget.PopupMenu
import android.widget.TextView

class LengthFragment : CalculatorFragment() {
    private lateinit var inputSymbol: TextView

    private lateinit var conversionResult: TextView
    private lateinit var conversionSymbolResult: TextView

    private lateinit var unit: Button
    private lateinit var options: Array<Button>

    private lateinit var lightYear: Button
    private lateinit var kilometer: Button
    private lateinit var hectometer: Button
    private lateinit var decameter: Button
    private lateinit var mile: Button
    private lateinit var meter: Button
    private lateinit var decimeter: Button
    private lateinit var centimeter: Button
    private lateinit var millimeter: Button
    private lateinit var micrometer: Button

    override fun onAttach(context: Context) {
        super.onAttach(context)
        this.context = context
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        fragmentView = inflater.inflate(R.layout.fragment_calculator_length, container, false)

        input = fragmentView.findViewById(R.id.input)
        inputSymbol = fragmentView.findViewById(R.id.input_symbol)

        input.isFocusable = false

        unit = fragmentView.findViewById(R.id.unit)

        conversionResult = fragmentView.findViewById(R.id.option_conversion_number_result)
        conversionSymbolResult = fragmentView.findViewById(R.id.option_conversion_symbol_result)

        val othersHorizontalView = fragmentView.findViewById<HorizontalScrollView>(R.id.others_horizontal_scroll_view)
        othersHorizontalView.isHorizontalScrollBarEnabled = false

        lightYear = fragmentView.findViewById(R.id.light_year)
        kilometer = fragmentView.findViewById(R.id.kilometer)
        hectometer = fragmentView.findViewById(R.id.hectometer)
        decameter = fragmentView.findViewById(R.id.decameter)
        mile = fragmentView.findViewById(R.id.mile)
        meter = fragmentView.findViewById(R.id.meter)
        decimeter = fragmentView.findViewById(R.id.decimeter)
        centimeter = fragmentView.findViewById(R.id.centimeter)
        millimeter = fragmentView.findViewById(R.id.millimeter)
        micrometer = fragmentView.findViewById(R.id.micrometer)

        options = arrayOf (
                lightYear,
                kilometer,
                hectometer,
                decameter,
                mile,
                meter,
                decimeter,
                centimeter,
                millimeter,
                micrometer
        )

        keypadButtons[10] = fragmentView.findViewById(R.id.decimal_separator)
        calculatorMode = fragmentView.findViewById(R.id.calculator_mode)
        delete = fragmentView.findViewById(R.id.delete)

        units()
        inputOption()

        // Default configuration (meter to kilometer).
        inputSymbol.text = getString(R.string.meter_symbol)
        (activity as MainActivity).selectButton(meter, options, R.drawable.option_clicked)
        (activity as MainActivity).preferencesEditor.putString("convertFrom", "meter")
                .apply()

        unit.text = getString(R.string.kilometer)
        conversionSymbolResult.text = getString(R.string.kilometer_symbol)
        (activity as MainActivity).preferencesEditor.putString("convertTo", "kilometer")
                .apply()

        (activity as MainActivity).calculatorMode(context, calculatorMode)

        (activity as MainActivity).inputNumber(fragmentView, input, conversionResult, conversionSymbolResult, input.text.toString())
        inputDecimalSeparator(input)
        (activity as MainActivity).delete(input, delete, conversionResult, conversionSymbolResult)

        return fragmentView
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun units() {
        unit.setOnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> (activity as MainActivity).bounceIn(unit, 0.5, 5.0)
                MotionEvent.ACTION_UP -> {
                    val units = PopupMenu(getContext(), unit)
                    val inflater: MenuInflater

                    inflater = units.menuInflater

                    inflater.inflate(R.menu.units_length, units.menu)
                    units.show()

                    units.setOnMenuItemClickListener { item ->
                        when {
                            item.title == getString(R.string.light_year) -> {
                                (activity as MainActivity).preferencesEditor.putString("convertTo", "lightYear")
                                        .apply()

                                conversionSymbolResult.text = getString(R.string.light_year_symbol)
                            }

                            item.title == getString(R.string.kilometer) -> {
                                (activity as MainActivity).preferencesEditor.putString("convertTo", "kilometer")
                                        .apply()

                                conversionSymbolResult.text = getString(R.string.kilometer_symbol)
                                unit.text = getString(R.string.kilometer)
                            }

                            item.title == getString(R.string.hectometer) -> {
                                (activity as MainActivity).preferencesEditor.putString("convertTo", "hectometer")
                                        .apply()

                                conversionSymbolResult.text = getString(R.string.hectometer_symbol)
                                unit.text = getString(R.string.hectometer)
                            }

                            item.title == getString(R.string.decameter) -> {
                                (activity as MainActivity).preferencesEditor.putString("convertTo", "decameter")
                                        .apply()

                                conversionSymbolResult.text = getString(R.string.decameter_symbol)
                                unit.text = getString(R.string.decameter)
                            }

                            item.title == getString(R.string.mile) -> {
                                (activity as MainActivity).preferencesEditor.putString("convertTo", "mile")
                                        .apply()

                                conversionSymbolResult.text = getString(R.string.mile_symbol)
                                unit.text = getString(R.string.mile)
                            }

                            item.title == getString(R.string.meter) -> {
                                (activity as MainActivity).preferencesEditor.putString("convertTo", "meter")
                                        .apply()

                                conversionSymbolResult.text = getString(R.string.meter_symbol)
                                unit.text = getString(R.string.meter)
                            }

                            item.title == getString(R.string.mile) -> {
                                (activity as MainActivity).preferencesEditor.putString("convertTo", "mile")
                                        .apply()

                                conversionSymbolResult.text = getString(R.string.mile_symbol)
                                unit.text = getString(R.string.mile)
                            }

                            item.title == getString(R.string.decimeter) -> {
                                (activity as MainActivity).preferencesEditor.putString("convertTo", "decimeter")
                                        .apply()

                                conversionSymbolResult.text = getString(R.string.decimeter_symbol)
                                unit.text = getString(R.string.decimeter)
                            }

                            item.title == getString(R.string.centimeter) -> {
                                (activity as MainActivity).preferencesEditor.putString("convertTo", "centimeter")
                                        .apply()

                                conversionSymbolResult.text = getString(R.string.centimeter_symbol)
                                unit.text = getString(R.string.centimeter)
                            }

                            item.title == getString(R.string.millimeter) -> {
                                (activity as MainActivity).preferencesEditor.putString("convertTo", "millimeter")
                                        .apply()

                                conversionSymbolResult.text = getString(R.string.millimeter_symbol)
                                unit.text = getString(R.string.millimeter)
                            }

                            item.title == getString(R.string.micrometer) -> {
                                (activity as MainActivity).preferencesEditor.putString("convertTo", "micrometer")
                                        .apply()

                                conversionSymbolResult.text = getString(R.string.micrometer_symbol)
                                unit.text = getString(R.string.micrometer)
                            }
                        }

                        (activity as MainActivity).calc(input, conversionResult, conversionSymbolResult)

                        true
                    }
                }
            }

            true
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun inputOption() {
        lightYear.setOnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> (activity as MainActivity).bounceIn(lightYear, 0.5, 5.0)
                MotionEvent.ACTION_UP -> {
                    lightYear.startAnimation((activity as MainActivity).bounceOut)

                    (activity as MainActivity).selectButton(lightYear, options, R.drawable.option_clicked)
                    inputSymbol.text = getString(R.string.light_year_symbol)

                    (activity as MainActivity).preferencesEditor.putString("convertFrom", "lightYear")
                            .apply()
                }
            }

            (activity as MainActivity).calc(input, conversionResult, conversionSymbolResult)

            true
        }

        kilometer.setOnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> (activity as MainActivity).bounceIn(kilometer, 0.5, 5.0)
                MotionEvent.ACTION_UP -> {
                    kilometer.startAnimation((activity as MainActivity).bounceOut)

                    (activity as MainActivity).selectButton(kilometer, options, R.drawable.option_clicked)
                    inputSymbol.text = getString(R.string.kilometer_symbol)

                    (activity as MainActivity).preferencesEditor.putString("convertFrom", "kilometer")
                            .apply()
                }
            }

            (activity as MainActivity).calc(input, conversionResult, conversionSymbolResult)

            true
        }

        hectometer.setOnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> (activity as MainActivity).bounceIn(hectometer, 0.5, 5.0)
                MotionEvent.ACTION_UP -> {
                    hectometer.startAnimation((activity as MainActivity).bounceOut)

                    (activity as MainActivity).selectButton(hectometer, options, R.drawable.option_clicked)
                    inputSymbol.text = getString(R.string.hectometer_symbol)

                    (activity as MainActivity).preferencesEditor.putString("convertFrom", "hectometer")
                            .apply()
                }
            }

            (activity as MainActivity).calc(input, conversionResult, conversionSymbolResult)

            true
        }

        decameter.setOnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> (activity as MainActivity).bounceIn(decameter, 0.5, 5.0)
                MotionEvent.ACTION_UP -> {
                    decameter.startAnimation((activity as MainActivity).bounceOut)

                    (activity as MainActivity).selectButton(decameter, options, R.drawable.option_clicked)
                    inputSymbol.text = getString(R.string.decameter_symbol)

                    (activity as MainActivity).preferencesEditor.putString("convertFrom", "decameter")
                            .apply()
                }
            }

            (activity as MainActivity).calc(input, conversionResult, conversionSymbolResult)

            true
        }

        mile.setOnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> (activity as MainActivity).bounceIn(mile, 0.5, 5.0)
                MotionEvent.ACTION_UP -> {
                    meter.startAnimation((activity as MainActivity).bounceOut)

                    (activity as MainActivity).selectButton(mile, options, R.drawable.option_clicked)
                    inputSymbol.text = getString(R.string.mile_symbol)

                    (activity as MainActivity).preferencesEditor.putString("convertFrom", "mile")
                            .apply()
                }
            }

            (activity as MainActivity).calc(input, conversionResult, conversionSymbolResult)

            true
        }

        meter.setOnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> (activity as MainActivity).bounceIn(meter, 0.5, 5.0)
                MotionEvent.ACTION_UP -> {
                    meter.startAnimation((activity as MainActivity).bounceOut)

                    (activity as MainActivity).selectButton(meter, options, R.drawable.option_clicked)
                    inputSymbol.text = getString(R.string.meter_symbol)

                    (activity as MainActivity).preferencesEditor.putString("convertFrom", "meter")
                            .apply()
                }
            }

            (activity as MainActivity).calc(input, conversionResult, conversionSymbolResult)

            true
        }

        decimeter.setOnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> (activity as MainActivity).bounceIn(decimeter, 0.5, 5.0)
                MotionEvent.ACTION_UP -> {
                    decimeter.startAnimation((activity as MainActivity).bounceOut)

                    (activity as MainActivity).selectButton(decimeter, options, R.drawable.option_clicked)
                    inputSymbol.text = getString(R.string.decimeter_symbol)

                    (activity as MainActivity).preferencesEditor.putString("convertFrom", "decimeter")
                            .apply()
                }
            }

            (activity as MainActivity).calc(input, conversionResult, conversionSymbolResult)

            true
        }

        centimeter.setOnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> (activity as MainActivity).bounceIn(centimeter, 0.5, 5.0)
                MotionEvent.ACTION_UP -> {
                    centimeter.startAnimation((activity as MainActivity).bounceOut)

                    (activity as MainActivity).selectButton(centimeter, options, R.drawable.option_clicked)
                    inputSymbol.text = getString(R.string.centimeter_symbol)

                    (activity as MainActivity).preferencesEditor.putString("convertFrom", "centimeter")
                            .apply()
                }
            }

            (activity as MainActivity).calc(input, conversionResult, conversionSymbolResult)

            true
        }

        millimeter.setOnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> (activity as MainActivity).bounceIn(millimeter, 0.5, 5.0)
                MotionEvent.ACTION_UP -> {
                    millimeter.startAnimation((activity as MainActivity).bounceOut)

                    (activity as MainActivity).selectButton(millimeter, options, R.drawable.option_clicked)
                    inputSymbol.text = getString(R.string.millimeter_symbol)

                    (activity as MainActivity).preferencesEditor.putString("convertFrom", "millimeter")
                            .apply()
                }
            }

            (activity as MainActivity).calc(input, conversionResult, conversionSymbolResult)

            true
        }

        micrometer.setOnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> (activity as MainActivity).bounceIn(micrometer, 0.5, 5.0)
                MotionEvent.ACTION_UP -> {
                    micrometer.startAnimation((activity as MainActivity).bounceOut)

                    (activity as MainActivity).selectButton(micrometer, options, R.drawable.option_clicked)
                    inputSymbol.text = getString(R.string.micrometer_symbol)

                    (activity as MainActivity).preferencesEditor.putString("convertFrom", "micrometer")
                            .apply()
                }
            }

            (activity as MainActivity).calc(input, conversionResult, conversionSymbolResult)

            true
        }
    }
}